File: 03085. Minimum Deletions to Make String K-Special/solution.go
/**
 * Problem: 3085. Minimum Deletions to Make String K-Special
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func minimumDeletions(word string, k int) int {
	// Count frequency of each character (a-z)
	charFreq := make([]int, 26)
	// Track cumulative deletions as we iterate
	cumulativeDeletions := 0
	// Initialize answer with worst case (delete all characters)
	minDeletions := len(word)

	// Count frequency of each character
	for _, ch := range word {
		charFreq[ch-'a']++
	}

	// Sort frequencies to process from lowest to highest
	sort.Ints(charFreq)

	// Try each frequency as the minimum frequency
	for i := 0; i < len(charFreq); i++ {
		// Start with current cumulative deletions
		currentDeletions := cumulativeDeletions
		// Set current frequency as minimum
		minFreq := charFreq[i]

		// Process frequencies from highest to lowest
		for j := len(charFreq) - 1; j > i; j-- {
			// If frequency difference is within k, no more deletions needed
			if charFreq[j]-minFreq <= k {
				break
			}

			// Delete excess characters beyond k difference
			currentDeletions += charFreq[j] - minFreq - k
		}
		
		// Update minimum deletions found so far
		minDeletions = min(minDeletions, currentDeletions)
		// Add current frequency to cumulative deletions for next iteration
		cumulativeDeletions += charFreq[i]
	}

	return minDeletions
}

File: 03085. Minimum Deletions to Make String K-Special/solution.js
/**
 * Problem: 3085. Minimum Deletions to Make String K-Special
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 10 ms (Beats 100%)
 */

/**
 * Calculates the minimum deletions to make character frequencies differ by at most k
 *
 * @param {string} word - Input string with lowercase letters
 * @param {number} k - Maximum allowed frequency difference
 *
 * @returns {number} - Minimum deletions needed
 */
const minimumDeletions = (word, k) => {
  // Count frequency of each character (a-z)
  const charFrequencies = new Int32Array(26)

  for (let i = 0; i < word.length; i++)
    charFrequencies[word.charCodeAt(i) - 97]++

  // Sort frequencies in ascending order
  charFrequencies.sort((a, b) => a - b)

  // Build suffix sum array for quick range sum queries
  const suffixSums = new Int32Array(charFrequencies.length + 1)

  for (let i = suffixSums.length - 2; i >= 0; i--)
    suffixSums[i] = (suffixSums[i + 1] || 0) + charFrequencies[i]

  // Track the rightmost valid position in sliding window
  let rightPointer = 0
  // Track minimum deletions found so far
  let minDeletions = suffixSums[0]
  // Track cumulative deletions for current position
  let currentDeletions = 0
  // Skip leading zeros in frequency array
  let startIndex = 0

  while (
    startIndex < charFrequencies.length &&
    charFrequencies[startIndex] === 0
  )
    startIndex++

  for (let i = startIndex; i < charFrequencies.length; i++) {
    // Expand window while frequency difference is within k
    while (
      rightPointer + 1 < suffixSums.length &&
      charFrequencies[rightPointer + 1] - charFrequencies[i] <= k
    )
      rightPointer++

    // Calculate deletions for current window configuration
    minDeletions = Math.min(
      minDeletions,
      suffixSums[rightPointer + 1] -
        (charFrequencies[i] + k) *
          (charFrequencies.length - (rightPointer + 1)) +
        currentDeletions
    )

    // Add current frequency to cumulative deletions
    currentDeletions += charFrequencies[i]
  }

  return minDeletions
}
